00X-ABI-Stable-Module-API.md

```diff
@@ -0,0 +1,179 @@
+
+# API definition
+
+**WORK IN PROGRESS, NOT ANYWHERE NEAR COMPLETE.** 
+
+This API will be built up from a minimal set as we explore the full set of Nan functions.  To start it will only include the minimal set that we need to get started.  What this means is we have not define the full API yet but instead want to give a feel of the general shape (which is not really quite there yet). 
+
+One of the main gaps is that we have not filled in the table for node::ni::New(ni_env,...) yet so we can not fully specify the arguments for the rest of the calls.  This will be one of the priorities as we work through the first Nan example.
+
+The proposal introduces the node::ni namespace to encapsulate all native module interface functions. ni_env is an opaque handle to a given instance of a node environment for use by a native module.  In the case of a C version of the interface ni_env would be the first paramter to all methods.  In the case of a C++ version all methods would be functions callable on an ni_env object.
+
+The definition of the API will be broken down into the following categories:
+
+## core functions
+
+### C setting javascript values
+<PRE>
+node::ni::SetMethod(ni_env, recv, name, callback)
+node::ni::SetPrototypeMethod(ni_env, recv, name, callback)
+node::ni::SetTemplate(ni_env,...)
+node::ni::SetPrototypeTemplate(ni_env,...)
+node::ni::SetInstanceTemplate(ni_env,...)
+</PRE>
+
+### C getting javascript values
+<PRE>
+node::ni::ObjectHandle node:ni::getGlobalObject(ni_env)
+node::ni::ObjectHandle::getProperty(ni_env,recv, ...)
+</PRE>
+
+### C calling javascript
+<PRE>
+node::ni::callMethod(ni_env,recv,,,)
+</PRE>
+
+## data types
+
+### Declarations
+<PRE>
+node::ni::New(ni_env,...)
+</PRE>
+
+New() takes a C type and returns a an opaque handle to the corresponding javascript
+object which is then passed into the API calls. Table will show the
+correspondance between the C types and the different types of opaque handles.
+
+| In  (C/C++)                            | out                                         |
+|----------------------------------------|---------------------------------------------|
+| char*                                  | PrimitiveStringHandle                       |
+| utf8*                                  | PrimitiveStringHandle                       |
+| u32                                    | PrimitiveU32                                |
+| ...                                    |                                             |
+(Table is obivously incomplete)
+
+
+### Argument Types
+
+<PRE>
+node::ni::FunctionCallbackInfo(ni_env,...)
+node::ni::PropertyCallbackInfo(ni_env,...)
+node::ni::ReturnValue(ni_env,...)
+Those created by New 
+</PRE>
+
+## object lifecycle
+
+<PRE>
+node::ni::PushScope(ni_env,...)
+node::ni::PopScope(ni_env,...)
+node::ni::CreatePersistent(ni_env,...)
+node::ni::FreePersistent(ni_env,...)
+node::ni::CreateGlobal(ni_env,...)
+node::ni::FreeGlobal(ni_env,...)
```

Comment:
Weak persistents (Finalizers) are quite necessary. node::ObjectWrap depends on them, as does node::Buffer, as do many addons, regardless of ObjectWrap or Buffers. If other engines cannot support them, how could they possibly support Node itself?

In regard to V8, Globals are movable, while Persistents are not. Globals are automatically reset when they are destroyed, Persistents are not (with the current default traits). But I do not know what the intent is regarding this API.